src/custom/state/orders/middleware.ts

```diff
@@ -0,0 +1,153 @@
+import { Middleware, isAnyOf } from '@reduxjs/toolkit'
+
+import * as OrderActions from './actions'
+import { OrderID } from 'utils/operator'
+import { addPopup } from 'state/application/actions'
+import { AppState } from 'state'
+
+const isSingleOrderChangeAction = isAnyOf(
+  OrderActions.addPendingOrder,
+  OrderActions.expireOrder,
+  OrderActions.fulfillOrder
+)
+
+const isPendingOrderAction = isAnyOf(OrderActions.addPendingOrder)
+
+const isFulfillOrderAction = isAnyOf(OrderActions.fulfillOrder)
+
+const isBatchOrderAction = isAnyOf(OrderActions.fulfillOrdersBatch, OrderActions.expireOrdersBatch)
+
+const isBatchFulfillOrderAction = isAnyOf(OrderActions.fulfillOrdersBatch)
+
+// what is passed to addPopup action
+type PopupPayload = Parameters<typeof addPopup>[0]
+interface OrderIDWithPopup {
+  id: OrderID
+  popup: PopupPayload
+}
+
+// on each Pending, Expired, Fulfilled order action
+// a corresponsing Popup action is dispatched
+export const popupMiddleware: Middleware<{}, AppState> = store => next => action => {
+  const result = next(action)
+
+  let idsAndPopups: OrderIDWithPopup[] = []
+  //  is it a singular action with {chainId, id} payload
+  if (isSingleOrderChangeAction(action)) {
+    const { id, chainId } = action.payload
+
+    // use current state to lookup orders' data
+    const orders = store.getState().orders[chainId]
+
+    if (!orders) return
+
+    const { pending, fulfilled, expired } = orders
+
+    const orderObject = pending?.[id] || fulfilled?.[id] || expired?.[id]
+
+    // look up Order.summary for Popup
+    const summary = orderObject?.order.summary
+
+    let popup: PopupPayload
+
+    if (isPendingOrderAction(action)) {
```

Comment:
This IF is huge, also the middleware code. Any chance of breaking this up, maybe aux functions?